### PR TITLE
Start time

### PR DESCRIPTION
**Pull request by dev-matan-tsuberi**
- One word keys
- Added `startTime` feature
- Updated & added some tests

### CODE REVIEW

File: contracts/Protected.sol [hunk 1]
```diff
@@ -69,19 +69,25 @@ contract Protected {
...
     {
         Key memory key = keys[_id][msg.sender];
         require(key.exists && isValidExpiration(key.expiration), "Invalid key");
         require(key.assignable, "Key is not assignable");
+        require(key.startTime == 0 || (_startTime >= key.startTime && _startTime > 0), "Cannot reduce key's start time");
         require(key.expiration == 0 || (_expiration <= key.expiration && _expiration > 0), "Cannot extend key's expiration");
-        require(key.uses == 0 || (_uses <= key.uses && _uses > 0), "Not enough uses avaiable");
+        require(_startTime == 0 || _expiration == 0 || _startTime < _expiration, "Start time must be strictly less than expiration");
         require(isValidExpiration(_expiration), "Expiration must be in the future");
+        require(_startTime == 0 || _startTime >= now, "Start time must be in the future");
```
Comment (ben-kaufman) [inline contracts/Protected.sol]: This condition doesn't make sense. Let's say I have a key with startTime of yesterday, now I want to transfer it without start time delay. This makes it impossible.

File: contracts/Protected.sol [hunk 2]
```diff
@@ -69,19 +69,25 @@ contract Protected {
...
+        uint80 _uses
     ) public
     {
         Key memory key = keys[_id][msg.sender];
         require(key.exists && isValidExpiration(key.expiration), "Invalid key");
         require(key.assignable, "Key is not assignable");
+        require(key.startTime == 0 || (_startTime >= key.startTime && _startTime > 0), "Cannot reduce key's start time");
         require(key.expiration == 0 || (_expiration <= key.expiration && _expiration > 0), "Cannot extend key's expiration");
-        require(key.uses == 0 || (_uses <= key.uses && _uses > 0), "Not enough uses avaiable");
+        require(_startTime == 0 || _expiration == 0 || _startTime < _expiration, "Start time must be strictly less than expiration");
```
Comment (ben-kaufman) [inline contracts/Protected.sol]: `_startTime == 0` requirement is unnecessary.

File: contracts/Protected.sol [hunk 3]
```diff
@@ -69,19 +69,25 @@ contract Protected {
...
         require(isValidExpiration(_expiration), "Expiration must be in the future");
+        require(_startTime == 0 || _startTime >= now, "Start time must be in the future");
+        require(key.uses == 0 || (_uses <= key.uses && _uses > 0), "Not enough uses avaiable");
 
+        bool possesKey = (!unlockable(_id, _to) && keys[_id][_to].startTime <= now);
```
Comment (ben-kaufman) [inline contracts/Protected.sol]: Why did you add `keys[_id][_to].startTime <= now`?
Also the `possesKey` name doesn't make sense as the condition means the opposite.

File: contracts/Protected.sol [hunk 4]
```diff
@@ -69,19 +69,25 @@ contract Protected {
...
         require(
-            !unlockable(_id, _to) || (keys[_id][_to].assignable == _assignable && keys[_id][_to].expiration == _expiration),
+            possesKey ||
+            (keys[_id][_to].assignable == _assignable && keys[_id][_to].expiration == _expiration && keys[_id][_to].startTime == _startTime),
```
Comment (ben-kaufman) [inline contracts/Protected.sol]: I think that a key can still be merged safely if both start times are in the past.

File: contracts/Protected.sol [hunk 5]
```diff
@@ -159,14 +167,18 @@ contract Protected {
...
         address _to,
         bool _assignable,
-        uint _expiration,
-        uint _uses
+        uint80 _startTime,
+        uint80 _expiration,
+        uint80 _uses
     ) internal
     {
+        require(_startTime == 0 || _expiration == 0 || _startTime < _expiration, "Start time must be strictly less than expiration");
```
Comment (ben-kaufman) [inline contracts/Protected.sol]: `_startTime == 0` requirement is unnecessary.

File: contracts/Protected.sol (hunk 1 above)
Comment (leviadam) [reply contracts/Protected.sol]: I agree. No reason _startTime should not be in the past.

**Comment by leviadam** [comment]
Deleted one of @ben-kaufman comments by mistake.
All in all looks good. Small fixes and we can approve this.

GG @dev-matan-tsuberi

File: contracts/Protected.sol (hunk 1 above)
Comment (dev-matan-tsuberi) [reply contracts/Protected.sol]: My thinking was that there's no point in having `startTime` yesterday because that time has passed. But yeah, it's unnecessary, and it means you can't use `grantFullKey` in that case.

**Comment by dev-matan-tsuberi** [comment]
Fix comments, added a couple of more tests.
@ben-kaufman Do you happen to remember what was the removed comment?

**Comment by dev-matan-tsuberi** [comment]
https://github.com/daostack/access_control/pull/35/commits/e554a49a9ac9ad10aa0db66045c38be8e64e489b: Resolve #36
